Katello/smart_proxy_container_gateway
# Fixes #37218 - podman login against the container registry returns 500 intermittently

Implementing improvements from https://github.com/Katello/smart_proxy_container_gateway/pull/32:

- Improves n+1 queries.
- Implements using a single DB connection across the entire plugin to solve DB race conditions.
- Changes the journaling mode in SQLite to WAL to gain performance benefits (https://www.sqlite.org/wal.html)
- Turns on cascading deletions for `repositories_users` when users or repositories are deleted.
  - This further simplifies queries.
- Adds transactions to avoid foreign key violation errors.

To test:
- All parts of the container gateway need to be exercised:
  - In Katello, create a mix of authenticated and unauthenticated repos
    - Achieve this by having at least one LCE with unauthenticated push `true`, and another with unauthenticated push `false`.
  - Try searching for the authenticated / unauthenticated repos. Ensure that `podman login` does change the results.
  - Try pulling the content as well.
- Performance must be tested next:
  - In one terminal, create a loop that runs `smart_proxy.sync_container_gateway` with at least a second in between runs
  - In multiple other terminals, log on and off the container gateway constantly.
  - Ensure that fewer errors occur than before. There may still be some depending on how many terminals are going. It should be better than before this change.
```
[root@centos8-proxy-devel ~]# cat /root/podman-login-spam 
#!/bin/bash

for i in {1..100}
do
	podman login `hostname` -u admin -p changeme
        podman logout `hostname`
done

```

Another thing to try would be concurrent logins:

```
#!/bin/bash
  
for i in {1..50}
do
        podman login `hostname` -u admin -p changeme &
done

wait
```
 
- Database migrations also need testing. Test the migration with an existing DB and with no DB. Ensure that there is some data in the database before running the migration with a DB.
